ambari-server/src/main/java/org/apache/ambari/server/serveraction/kerberos/CreateKeytabFilesServerAction.java

```diff
@@ -231,6 +242,9 @@ protected CommandReport processIdentity(ResolvedKerberosPrincipal resolvedPrinci
                       commandReport = createCommandReport(1, HostRoleStatus.FAILED, "{}", actionLog.getStdOut(), actionLog.getStdErr());
                     } else {
                       try {
+                        if(regenerateKeytabs) {
+                          Keytab keytab = createKeytab(resolvedPrincipal.getPrincipal(), password, keyNumber, operationHandler, visitedPrincipalKeys != null, true, actionLog);
```

Comment:
I guess the proper place to do the change is `CreatePrincipalsServerAction`, there is a place like: 
```
      } else if (!StringUtils.isEmpty(kerberosPrincipalEntity.getCachedKeytabPath())) {
        // This principal has been processed and a keytab file has been cached for it... do not process it.
        processPrincipal = false;
      } else {
```
if we change it to something like: 
```
      } else if (!StringUtils.isEmpty(kerberosPrincipalEntity.getCachedKeytabPath())) {
        File file = new File(kerberosPrincipalEntity.getCachedKeytabPath());

        // Process the principal if the cache is missing, otherwise skip it
        processPrincipal = !file.exists();
      } else {
```
It would process with the the proper keytab re-generation

Reply:
@hapylestat I thought changing CreateKeytab operation code is more relevant - in the above recommendation we are touching the CreatePrincipleAction right.

Reply:
> we can't create here a keytab as `password` would be always `null` and keytab would be invalid at the end

Look like we don't need password here - see https://github.com/apache/ambari/blob/branch-2.7/ambari-server/src/main/java/org/apache/ambari/server/serveraction/kerberos/KDCKerberosOperationHandler.java#L177

Reply:
for Kerberos yes, but check please code for non-Kerberos handlers